File: src/shared/chatStreamBatch.ts
/**
 * Coarse chat stream flushes — thresholds from resources/contracts/chatStreamBatch.json.
 * Token deltas accumulate; UI updates on size, time, or paragraph boundaries.
 */

import contract from "../../resources/contracts/chatStreamBatch.json";

export interface ChatStreamBatchPolicy {
  minChars: number;
  maxHoldMs: number;
  newlineMinChars: number;
}

export const CHAT_STREAM_BATCH: ChatStreamBatchPolicy = {
  minChars: contract.minChars,
  maxHoldMs: contract.maxHoldMs,
  newlineMinChars: contract.newlineMinChars,
};

export type StreamBatchClock = () => number;

/** Pure batcher for assistant/note text UI flushes. */
export class StreamTextBatcher {
  private buffer = "";
  private bufferStartedAt: number | null = null;
  private readonly now: StreamBatchClock;

  constructor(
    private readonly policy: ChatStreamBatchPolicy = CHAT_STREAM_BATCH,
    now: StreamBatchClock = () => Date.now()
  ) {
    this.now = now;
  }

  push(chunk: string): string | null {
    if (!chunk) return null;
    if (this.buffer.length === 0) this.bufferStartedAt = this.now();
    this.buffer += chunk;
    return this.maybeFlush(false);
  }

  /** Force remaining buffer out (stream end, tool boundary). */
  flush(): string | null {
    return this.maybeFlush(true);
  }

  get pending(): string {
    return this.buffer;
  }

  private maybeFlush(force: boolean): string | null {
    if (!this.buffer) return null;
    if (force) return this.takeAll();

    const charLen = [...this.buffer].length;
    const heldLong =
      this.bufferStartedAt != null && this.now() - this.bufferStartedAt >= this.policy.maxHoldMs;

    const lastNl = this.buffer.lastIndexOf("\n");
    if (lastNl >= 0) {
      const prefix = this.buffer.slice(0, lastNl + 1);
      const prefixChars = [...prefix].length;
      if (prefixChars >= this.policy.newlineMinChars) {
        this.buffer = this.buffer.slice(lastNl + 1);
        this.bufferStartedAt = this.buffer ? this.now() : null;
        return prefix;
      }
    }

    if (charLen >= this.policy.minChars || heldLong) {
      return this.takeAll();
    }
    return null;
  }

  private takeAll(): string {
    const out = this.buffer;
    this.buffer = "";
    this.bufferStartedAt = null;
    return out;
  }
}

File: ios/HarnessMobile/Chat/ChatStreamBatch.swift
import Foundation

/// Coarse chat stream flushes — thresholds from bundled `resources/contracts/chatStreamBatch.json`.
struct ChatStreamBatchPolicy: Equatable {
    var minChars: Int
    var maxHoldMs: Int
    var newlineMinChars: Int

    static let defaults = ChatStreamBatchPolicy.loadContractDefaults()

    private static func loadContractDefaults() -> ChatStreamBatchPolicy {
        guard let url = Bundle.main.url(forResource: "chatStreamBatch", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let minChars = json["minChars"] as? Int, minChars > 0,
              let maxHoldMs = json["maxHoldMs"] as? Int, maxHoldMs > 0,
              let newlineMinChars = json["newlineMinChars"] as? Int, newlineMinChars > 0
        else {
            assertionFailure("resources/contracts/chatStreamBatch.json failed to load or parse from the app bundle")
            return ChatStreamBatchPolicy(minChars: 150, maxHoldMs: 350, newlineMinChars: 40)
        }
        return ChatStreamBatchPolicy(
            minChars: minChars,
            maxHoldMs: maxHoldMs,
            newlineMinChars: newlineMinChars
        )
    }
}

/// Accumulates token deltas and yields larger UI chunks.
struct StreamTextBatcher {
    private var buffer = ""
    private var bufferStartedAt: ContinuousClock.Instant?
    private let policy: ChatStreamBatchPolicy
    private let clock: ContinuousClock

    init(policy: ChatStreamBatchPolicy = .defaults, clock: ContinuousClock = ContinuousClock()) {
        self.policy = policy
        self.clock = clock
    }

    mutating func push(_ chunk: String) -> String? {
        guard !chunk.isEmpty else { return nil }
        if buffer.isEmpty {
            bufferStartedAt = clock.now
        }
        buffer += chunk
        return maybeFlush(force: false)
    }

    mutating func flush() -> String? {
        maybeFlush(force: true)
    }

    var pending: String { buffer }

    private mutating func maybeFlush(force: Bool) -> String? {
        guard !buffer.isEmpty else { return nil }
        if force { return takeAll() }

        let charLen = buffer.count
        let heldLong: Bool = {
            guard let started = bufferStartedAt else { return false }
            return clock.now - started >= .milliseconds(policy.maxHoldMs)
        }()

        if let nl = buffer.lastIndex(of: "\n") {
            let prefix = String(buffer[...nl])
            if prefix.count >= policy.newlineMinChars {
                buffer = String(buffer[buffer.index(after: nl)...])
                bufferStartedAt = buffer.isEmpty ? nil : clock.now
                return prefix
            }
        }

        if charLen >= policy.minChars || heldLong {
            return takeAll()
        }
        return nil
    }

    private mutating func takeAll() -> String {
        let out = buffer
        buffer = ""
        bufferStartedAt = nil
        return out
    }
}
